README.md
# android-icons
[![Build Status](https://travis-ci.com/randytarampi/android-icons.svg?branch=master)](https://travis-ci.com/randytarampi/android-icons)
[![Dependency Status](https://david-dm.org/randytarampi/android-icons.svg)](https://david-dm.org/randytarampi/android-icons)
[![devDependency Status](https://david-dm.org/randytarampi/android-icons/dev-status.svg)](https://david-dm.org/randytarampi/android-icons#info=devDependencies)
[![Semantically Released](https://img.shields.io/badge/versioning-semantically%20released-brightgreen.svg)](https://github.com/boennemann/semantic-release) 

> Get android icon file names and dimensions (width)

The default icon file names (as used by [cordova](https://cordova.apache.org/docs/en/4.0.0/config_ref_images.md.html#Icons%20and%20Splash%20Screens)) and required sizes for android are listed in a [JSON file](icons.json). This information is useful, for example, when you want to generate icons with the required sizes or to create a [`config.xml`](http://docs.phonegap.com/en/3.5.0/config_ref_images.md.html) file for a PhoneGap/Cordova project or if you just need to create the icons for your android project from one source icon.
This information also exists for [iOS icons](https://github.com/randytarampi/ios-icons).


## Install

```sh
$ npm install --save @randy.tarampi/android-icons
```


## Usage

```js
var icons = require('@randy.tarampi/android-icons');

icons();
//=> [{ "name": "mdpi.png", "width": 48 }, { "name": "hdpi.png", "width": 72 }, ... ]

icons({size: '48'})
//=> {name: 'mdpi.png', width: 48}

icons({size: 'xhdpi'})
//=> {name: 'xhdpi.png', width: 96}
```


## API

### icons()

Returns an array of icons, each icon being represented by an object with `name` and `width` properties.

> Notice that icons are always squares, so no `height` property is provided.

### icons(options)
#### options

Only option for now is `size`, which can be either a `Number` or `String` value. If it is a `Number`, it represents the width in pixels. If it is a `String`, you can use `"mdpi"`, `"hdpi"` etc. to refer to a certain size or the complete file name, e.g. `xhdpi.png`.

Returns icon object for that size or `null`.

For example:

```js
icons({size: 'xhdpi'})
//=> {name: 'xhdpi.png', width: 96}
```


## CLI
> android-icons logs to stdout in comma-separated values format (csv) by default so you can easy pipe to other commands in UNIX systems.

```sh
$ npm install --global @randy.tarampi/android-icons
```

```sh
$ android-icons --help
Get android icon file names and dimensions (width)

Use `--format json` to set output to JSON.
Get specifc icon by size or name by using `--size`.

Examples:
  $ android-icons --size 48
  icon-40@2x.png,80

  $ android-icons --size 80 --format json
  {"name":"icon-40@2x.png","width":80}

  $ android-icons --size small
  icon-small.png,29
```

## Semantic Releases
This module is being [semantically released](https://github.com/boennemann/semantic-release). You can safely use `"^1.0.0"` in your `package.json`.

## License
MIT © [David Pfahler](http://excellenteasy.com)


CHANGELOG.md
## [3.1.2](https://github.com/randytarampi/android-icons/compare/v3.1.1...v3.1.2) (2020-04-30)


### Bug Fixes

* upgrade @semantic-release/changelog from 5.0.0 to 5.0.1 ([6924876](https://github.com/randytarampi/android-icons/commit/6924876219e1ea7db67fa27bfa0b7b6a2d010e68))
* upgrade @semantic-release/changelog from 5.0.0 to 5.0.1 ([f025f0b](https://github.com/randytarampi/android-icons/commit/f025f0bd6c79ddc7b89784b0a8e4e087f8fd40f9))
* upgrade semantic-release from 17.0.0 to 17.0.7 ([59fccc1](https://github.com/randytarampi/android-icons/commit/59fccc116ec5d0384c25f7412478265a2ff60182))
* upgrade semantic-release from 17.0.0 to 17.0.7 ([edaea41](https://github.com/randytarampi/android-icons/commit/edaea411afca1a4affcb2e7769d4cf7cf01bfe20))
* upgrade standard from 14.3.1 to 14.3.3 ([ee84324](https://github.com/randytarampi/android-icons/commit/ee84324fc133e5908597a750313b43996e04cda4))
* upgrade standard from 14.3.1 to 14.3.3 ([b6318f8](https://github.com/randytarampi/android-icons/commit/b6318f84ed749eff7021d980276472ff1ec67b85))
* upgrade yargs from 15.3.0 to 15.3.1 ([bafa831](https://github.com/randytarampi/android-icons/commit/bafa83117d82aba783d9d99e2e3a19059006fd0e))
* upgrade yargs from 15.3.0 to 15.3.1 ([9df01ed](https://github.com/randytarampi/android-icons/commit/9df01ed6b6e792dbea3ca74ef2fee8e67cc952e8))

## [3.1.1](https://github.com/randytarampi/android-icons/compare/v3.1.0...v3.1.1) (2020-04-19)


### Bug Fixes

* package.json & package-lock.json to reduce vulnerabilities ([e9a1328](https://github.com/randytarampi/android-icons/commit/e9a1328a725a883405ae546a73494b52492055fc))

# [3.1.0](https://github.com/randytarampi/android-icons/compare/v3.0.0...v3.1.0) (2020-01-06)


### Features

* **package:** Blindly upgrade our dependencies for the new year and support node@10, node@12 and beyond. ([269473e](https://github.com/randytarampi/android-icons/commit/269473e950ee8f479a7fa2c2af0c035db7c65ca4))

# [3.0.0](https://github.com/randytarampi/android-icons/compare/v2.1.0...v3.0.0) (2019-03-10)


### Bug Fixes

* **package:** Since the package updates we no longer support node@4. ([17bea94](https://github.com/randytarampi/android-icons/commit/17bea94))
* **travis:** Don't bother caching dependencies. ([fcc5cf4](https://github.com/randytarampi/android-icons/commit/fcc5cf4))


### BREAKING CHANGES

* **package:** Drop support for node 4 and earlier, since `yargs^13.2.2` requires ES6 support.

# [2.1.0](https://github.com/randytarampi/android-icons/compare/v2.0.4...v2.1.0) (2019-03-10)


### Features

* Add PWA splash screen icons. ([2ee286c](https://github.com/randytarampi/android-icons/commit/2ee286c)), closes [/github.com/GoogleChrome/lighthouse/issues/291#issue-153328672](https://github.com//github.com/GoogleChrome/lighthouse/issues/291/issues/issue-153328672)
* **travis:** `android-icons` -> `[@randy](https://github.com/randy).tarampi/android-icons` ([52db43a](https://github.com/randytarampi/android-icons/commit/52db43a))


test.js
'use strict'
var test = require('tape')
var androidIcons = require('./')
var exec = require('child_process').exec

test('returns all icons in array', function (t) {
  t.plan(2)
  var icons = androidIcons()
  t.ok(Array.isArray(icons), 'returned an array')
  t.equal(icons.length, 9, '9 icons returned')
})

test('returns icon for size 48 as Number', function (t) {
  t.plan(2)
  var icon = androidIcons({ size: 48 })
  t.ok(icon.name === 'mdpi.png', 'icon name correct')
  t.ok(icon.width === 48, 'icon width correct')
})

test('returns icon for size 48 as String', function (t) {
  t.plan(2)
  var icon = androidIcons({ size: '48' })
  t.ok(icon.name === 'mdpi.png', 'icon name correct')
  t.ok(icon.width === 48, 'icon width correct')
})

test('returns icon for size 96', function (t) {
  t.plan(2)
  var icon = androidIcons({ size: 96 })
  t.ok(icon.name === 'xhdpi.png', 'icon name correct')
  t.ok(icon.width === 96, 'icon width correct')
})

test('returns icon for size xhdpi', function (t) {
  t.plan(2)
  var icon = androidIcons({ size: 'xhdpi' })
  t.ok(icon.name === 'xhdpi.png', 'icon name correct')
  t.ok(icon.width === 96, 'icon width correct')
})

test('cli returns all icons as csv', function (t) {
  t.plan(1)
  var expected = [
    'mdpi.png,48',
    'hdpi.png,72',
    'xhdpi.png,96',
    'xxhdpi.png,144',
    'xxxhdpi.png,192',
    'splash-pwa-hdpi.png,192',
    'splash-pwa-xhdpi.png,256',
    'splash-pwa-xxhdpi.png,384',
    'splash-pwa-xxxhdpi.png,512\n'
  ].join('\n')
  exec('./bin/android-icons.js', function (error, stdout, stderr) {
    var err = error || stderr
    if (err) {
      return t.fail('calling cli produced an error: ' + err)
    }
    t.equal(stdout, expected, 'cli returned expected output')
  })
})

test('cli returns icon for size hdpi', function (t) {
  t.plan(1)
  var expected = 'hdpi.png,72\n'
  exec('./bin/android-icons.js --size 72', function (error, stdout, stderr) {
    var err = error || stderr
    if (err) {
      return t.fail('calling cli produced an error: ' + err)
    }
    t.equal(stdout, expected, 'cli returned expected output')
  })
})

test('cli returns icon for size xhdpi', function (t) {
  t.plan(1)
  var expected = 'xhdpi.png,96\n'
  exec('./bin/android-icons.js --size xhdpi', function (error, stdout, stderr) {
    var err = error || stderr
    if (err) {
      return t.fail('calling cli produced an error: ' + err)
    }
    t.equal(stdout, expected, 'cli returned expected output')
  })
})

test('cli returns all icons as json w/abbreviated flags', function (t) {
  t.plan(3)
  var expected = '[{"name":"mdpi.png","width":48},{"name":"hdpi.png","width":72},{"name":"xhdpi.png","width":96},{"name":"xxhdpi.png","width":144},{"name":"xxxhdpi.png","width":192},{"name":"splash-pwa-hdpi.png","width":192},{"name":"splash-pwa-xhdpi.png","width":256},{"name":"splash-pwa-xxhdpi.png","width":384},{"name":"splash-pwa-xxxhdpi.png","width":512}]\n'
  exec('./bin/android-icons.js --forma json', function (error, stdout, stderr) {
    var err = error || stderr
    if (err) {
      return t.fail('calling cli produced an error: ' + err)
    }
    t.equal(stdout, expected, 'cli returned expected output')
  })
  exec('./bin/android-icons.js --for json', function (error, stdout, stderr) {
    var err = error || stderr
    if (err) {
      return t.fail('calling cli produced an error: ' + err)
    }
    t.equal(stdout, expected, 'cli returned expected output')
  })
  exec('./bin/android-icons.js --f json', function (error, stdout, stderr) {
    var err = error || stderr
    if (err) {
      return t.fail('calling cli produced an error: ' + err)
    }
    t.equal(stdout, expected, 'cli returned expected output')
  })
})

test('cli returns icon for size w/abbreviated flag', function (t) {
  t.plan(3)
  var expected = 'hdpi.png,72\n'
  exec('./bin/android-icons.js --siz 72', function (error, stdout, stderr) {
    var err = error || stderr
    if (err) {
      return t.fail('calling cli produced an error: ' + err)
    }
    t.equal(stdout, expected, 'cli returned expected output')
  })
  exec('./bin/android-icons.js --si 72', function (error, stdout, stderr) {
    var err = error || stderr
    if (err) {
      return t.fail('calling cli produced an error: ' + err)
    }
    t.equal(stdout, expected, 'cli returned expected output')
  })
  exec('./bin/android-icons.js --s 72', function (error, stdout, stderr) {
    var err = error || stderr
    if (err) {
      return t.fail('calling cli produced an error: ' + err)
    }
    t.equal(stdout, expected, 'cli returned expected output')
  })
})

test('cli returns icon for size xhdpi w/ abbreviated flag', function (t) {
  t.plan(1)
  var expected = 'xhdpi.png,96\n'
  exec('./bin/android-icons.js --size xhdpi', function (error, stdout, stderr) {
    var err = error || stderr
    if (err) {
      return t.fail('calling cli produced an error: ' + err)
    }
    t.equal(stdout, expected, 'cli returned expected output')
  })
})


index.js
'use strict'
var icons = require('./icons.json')
var idRegEx = /^(.*)\.png$/
var widths = icons.map(function (icon) {
  return icon.width
})
var ids = icons.map(function (icon) {
  return icon.name.match(idRegEx)[1]
})

function getIconForSize (size) {
  var width = getWidthForSize(size)

  if (!width) {
    return null
  }

  return icons[widths.indexOf(width)] || null
}

function getWidthForSize (size) {
  if (typeof size === 'number') return size

  var width = Number(size)
  if (!isNaN(width)) return width

  width = widths[ids.indexOf(size)]
  if (width) return width

  return widths[ids.indexOf(size.match(idRegEx)[1])]
}

module.exports = function (options) {
  options = options || {}
  var size = options.size
  if (!size) return icons

  return getIconForSize(size)
}

module.exports.icons = icons
module.exports.getIconForSize = getIconForSize
module.exports.getWidthForSize = getWidthForSize


cli.js
#!/usr/bin/env node
'use strict'
var abbrev = require('abbrev')
var argv = require('minimist')(process.argv.slice(2), abbrev('help', 'version', 'size', 'format'))
var pkg = require('./package.json')
var icons = require('./')

function help () {
  console.log([
    pkg.description,
    '',
    'Use `--format json` to set output to JSON.',
    'Get specifc icon by size or name by using `--size`.',
    '',
    'Examples:',
    '  $ android-icons --size 48',
    '  mdpi.png,48',
    '',
    '  $ android-icons --size 48 --format json',
    '  {"name":"mdpi.png","width":48}',
    '',
    '  $ android-icons --size xhdpi',
    '  xhdpi.png,96'
  ].join('\n'))
}

function formatLog (icons, argv) {
  var format = (argv.format || 'csv').toLowerCase()
  if (format === 'json') {
    return JSON.stringify(icons)
  }
  if (!Array.isArray(icons)) {
    icons = [icons]
  }
  return icons.map(function (icon) {
    return icon.name + ',' + icon.width
  }).join('\n')
}

function cli () {
  if (argv.help) return help()

  if (argv.version) return console.log(pkg.version)

  var options = {
    size: argv.size || argv.s
  }

  var output = icons(options)
  if (output) console.log(formatLog(output, argv))
}

cli()
